### DISCUSSION TITLE
How to stop the Keycloak server with Quarkus ?

### DISCUSSION THREAD

**Discussion by Davidgiordano74** [keycloak/keycloak#13090] [Archive]
Hello,

I try to find how to stop Keycloak with a command line (and not a Ctrl+C command). It's easy to start the server with `kc.sh start` but how to stop it ?

When I was on Keycloak 16, I used the command `jboss-cli.sh --connect command=:shutdown` but of course jboss-cli doesn't exist anymore with Quarkus.

I saw we can save the pid in a file when we start the server and do a "cat" on it to kill it, but I hope for a cleaner solution.

Thanks for your help.

**Comment by AndreasGeisler** [comment]
Hello, I would also like to see a proper shutdown method for Keycloak. Currently on Windows, I need to use `taskkill -f -t -pid <PID of kc.bat>`.

**Comment by soorajjis** [comment]
Hello, I would also like to see a way to shutdown keycloak(quarkus based) gracefully.

**Comment by Ekouyoja** [comment]
Hello,
i have build a service on my server. Don't know if it's the best way. But i simply do systemctl stop keycloak that send a sigterm to the process.
[keycloak.service.txt](https://github.com/keycloak/keycloak/files/10040949/keycloak.service.txt)

**Reply by soorajjis** [reply]
Thanks Ekouyoja, but I am looking for windows and for graceful shutdown.

**Comment by sbacquet** [comment]
I'm looking for the same thing. Seems incredible there is no official way to gracefully stop Keycloak...

**Reply by brian-farrell** [reply]
Agreed - this is absolute madness that there is no way to gracefully shutdown the server.   And if there is a way, it's absolutely crazy that there is no easily discovered documentation on how to do so.

**Comment by marlon-sousa** [comment]
Hello,

My request is similar, but I need to exit programatically from inside an SPI.

of course system.exit() works, but then all my logging in jboss logger is not writteen (I have tried to slush it but I have found no method to do it), and using keycloakPlatform exit method does nothing.

Any idea of what is the recomended way to gracefully shutdown KC from inside an SPI?

**Reply by danielFesenmeyer** [reply]
I think there is no recommended way, because it's something you normally should just not do.
But maybe you could try with `Quarkus.asyncExit` (see https://quarkus.io/guides/command-mode-reference).

**Comment by marlon-sousa** [comment]
Hello,

I have found this, but I haven't found a way of accessing the Quarkus  object from within ,my spi.

I thought this is what the Platform would do for me, but obviously I missed something.

Do you know how can I access this object from my SPI?

**Reply by danielFesenmeyer** [reply]
I think you can add a provided dependency to `io.quarkus:quarkus-core`, with the quarkus version currently used by Keycloak.

**Comment by dkrysiak** [comment]
[My StackOverflow answer](https://stackoverflow.com/a/79640255/3673367) (based on some other one): my preferred solution is:
Start:
```shell
./keycloak-21.1.2/bin/kc.sh start-dev --http-port=8080 > keycloak.stdout 2>&1
echo -n "$!" > keycloak.pid
```
Stop:
```shell
$ cat keycloak.pid | xargs kill -INT
```
Use `-INT` instead of `-TERM` because then Keycloak has a chance to handle the request to stop. In `keycloak.stdout` I see several messages that my custom providers' `close()` method was called, and, in the end:
> INFO  [io.quarkus] (Shutdown thread) Keycloak stopped in 0.172s

So it looks that Keycloak was actually gracefully closed.
